spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -164,15 +178,81 @@ private String createMessagingErrorMessage(String description, Object payload) {
 	/**
 	 * Delegates payload extraction to
 	 * {@link #extractMessage(org.springframework.amqp.core.Message message)}
-	 * to enforce backward compatibility.
+	 * to enforce backward compatibility. Uses this listener adapter's converter instead of
+	 * the one configured in the converter adapter.
+	 * If the inbound message has no type information and the configured message converter
+	 * supports it, we attempt to infer the conversion type from the method signature.
 	 */
 	private class MessagingMessageConverterAdapter extends MessagingMessageConverter {
 
+		private final Object bean;
+
+		private final Method method;
+
+		private final Type inferredArgumentType;
+
+		public MessagingMessageConverterAdapter(Object bean, Method method) {
+			this.bean = bean;
+			this.method = method;
+			this.inferredArgumentType = determineInferredType();
+		}
+
 		@Override
 		protected Object extractPayload(org.springframework.amqp.core.Message message) {
+			MessageProperties messageProperties = message.getMessageProperties();
+			if (this.bean != null) {
+				messageProperties.setTargetBean(this.bean);
+			}
+			if (this.method != null) {
+				messageProperties.setTargetMethod(this.method);
+				if (this.inferredArgumentType != null) {
+					message.getMessageProperties().setInferredArgumentType(this.inferredArgumentType);
+				}
+			}
 			return extractMessage(message);
 		}
 
+		private Type determineInferredType() {
+			if (this.method == null) {
+				return null;
+			}
+			Annotation[][] parameterAnnotations = this.method.getParameterAnnotations();
+			// Single param; no annotation or @Payload
+			Type genericParameterType = null;
+			if (parameterAnnotations.length == 1) {
+				MethodParameter methodParameter = new MethodParameter(this.method, 0);
+				if (methodParameter.getParameterAnnotations().length == 0
+						|| methodParameter.hasParameterAnnotation(Payload.class)) {
+					genericParameterType = methodParameter.getGenericParameterType();
+				}
+			}
+			if (genericParameterType == null) {
+				for (int i = 0; i < parameterAnnotations.length; i++) {
+					MethodParameter methodParameter = new MethodParameter(this.method, i);
+					/*
+					 * We're looking for a single non-annotated parameter, or one annotated with
+					 * @Payload.
+					 * We ignore parameters with type Message because they are not involved with
+					 * conversion.
```

Comment:
OK. Why don't you ignore `Message` for the single parameter then?
And I guess that should be independently of the `@Payload` on it.

Right, you have that below, so, let's copy the logic to the single param!